https://issues.redhat.com/browse/ACM-17497 - Add Architecture section to Edge Manager (#7459)
open-cluster-management/rhacm-docs: edge_management/edge_mgmt_arch.adoc

[#edge-mgmt-arch]
= Understanding the {rhem} architecture

> The {rhem} uses an agent-based architecture with image-based operating systems (OS), that allows for a scalable and robust device management even under limited network conditions. Manage individual devices or an entire fleet by using the {rhem}.

Reviewer: ```suggestion
The {rhem} uses an agent-based architecture with image-based operating systems that allow for a scalable and robust device management, even with limited network conditions. Manage individual devices or an entire fleet by using the {rhem}.
```

Author: I think it should be `allows` here because the subject is the [agent-based architecture that utilizes the image-based operating systems].

> The {rhem} uses an agent-based architecture with image-based operating systems (OS), that allows for a scalable and robust device management even under limited network conditions. Manage individual devices or an entire fleet by using the {rhem}.

Reviewer 2: I would keep the aspect of "agent-based architecture" and "image-based OS" separate; they are complementary but not dependent. For example:

```suggestion
The {rhem} uses an agent-based architecture that allows for a scalable and robust device management even under limited network conditions. By deploying a {rhem} agent to a devices, the agent cana autonomously manage and monitor the device while periodically communicating with the {rhem} service to check for new configurations and to report device status.

The {rhem} matches particularly well with image-based OSes, as the {rhem} agent and its configuration can be included in the image when it is distributed to the devices. Image-based OSes further allow the agent to initiate a transactional update of the OS image and rollback to the previous version in case of an update error.
```

> The image-based OS that is distributed to the devices includes a {rhem} agent and its configuration. This means that the {rhem} agent autonomously manages and monitors each device while periodically communicating with the {rhem} service to check for new configurations and to report device status.

Reviewer: [operating system (noun)](https://redhat-documentation.github.io/supplementary-style-guide/#operating-system)
Incorrect forms: OS, Operating System
Source is the [Word usage](https://redhat-documentation.github.io/supplementary-style-guide/#_o) in the IBM Style Manual.

> The {rhem} architecture has the following main concepts:

Reviewer: Can we change "concepts" here? (this is more for an "idea" or in our role, a topic with info but no procedure)

Recommending "features","functions"??

Reviewer 2: The upstream docs use "concepts" in the meaning of "idea": Concepts the user should understand and be familiar with, because they are fundamental to how RHEM works and used throughout the docs.

Reviewer: I don't agree that we would use _concepts_ here, though. Upstream doc and our standards are going to be different...expect many changes when we write. We try not to make assumptions in our content. Definition for concepts:

https://www.merriam-webster.com/dictionary/concept

Reviewer: Usually in architecture doc, we list our features of a technology, or function.

* Agent
* Service
* Image-based OS
* API server 
* Database
* Device
* Device fleet

> For more information, see the following documentation:

Reviewer: Learn more from the following sections: (only because when we say "documentation" I was thinking we were linking to other files in the doc.)

> * <<agent-service,Understanding the {rhem} agent and service>>

Reviewer: We can remove _Understanding_ here. They are reading to understand, so it's extra.

* <<rhem-api-server,{rhem} API server>>
////
TODO include links when docs are ready
- Build images
- Provisioning
- Managing devices
- Managing device fleets
////

[#rhem-agent-service]
== {rhem} agent and service

> The {rhem} agent is a process running on each managed device that periodically communicates with the {rhem} service that runs on the {acm-short} hub cluster.

Reviewer: this is a mouthful. Can we shorten a bit?
```suggestion
The {rhem} agent runs on each managed device that periodically communicates with the {rhem} service that runs on the {acm-short} hub cluster.
```

Reviewer: I am not sure what communicates here, tho...

is it the process or the device that communicates...seems like device, but if that is not the case, we should see what we can do to reword this.

Reviewer 2: It's the agent that communicates with the service. It's the service that runs on the hub cluster.

Reviewer 2: It may be too colloquial, but it is very common to say that "the agent calls home to the service" instead of "communicates with". It is an expression that immediately clarifies it is the _agent_ that initiates the communication, which is an important property that addresses a lot of networking challenges. Something we could consider elaborating on here.

During periodic communication, the agent enrolls the devices into the service, checks for changes, and reports device status.

> When the agent is not actively connected to the service, the agent can apply configuration updates independently from the service.

Reviewer 2: Nees a bit of untangling:

- The agent enrolls the device only once at the very beginning (not periodically).
- The agent periodically checks with the service whether the service has updated the device's specification of the OS, configuration, applications, etc. to run.
- After learning about the updated specification, the agent can then apply the necessary updates independently from the service.

Reviewer: We have to keep the content within style constraints, so saying things like "the agent calls" -- this is considered an "anthropomorphism". The content is translated and written for accessibility. Much of the edits that you see us make is due to the industry and IBM/RH standards that writers should be using for content--it is required.

The {rhem} service handles user and agent authentication and authorization, device enrollment and inventory, rolling out updates to devices, and reporting status from individual devices or fleets.

> The service communicates with an SQL database that stores the inventory and runtime information, such as the current target configuration and the reported actual configuration.

Reviewer: Consider a list for line 37?

*Note:* You can maintain fleet definitions in Git that the {rhem} periodically syncs with the fleet definitions in the SQL database.

> 

Reviewer: Line 40 is a hard one.

You can maintain fleet definitions in Git. The {rhem} periodically syncs with Git and the fleet definitions in the SQL database.

??

Reviewer 2: I would not mention "SQL", because that is an implementation detail most users do not need to know.

The key point is that the service keeps in its database an inventory of enrolled devices and their "device specifications". The specification describes the configuration the user _intends_ the device to have. The service and agent work together to make that intend reality. The service also keeps a list of "fleet specifications" that describe the user intent for a whole fleet of devices.

The user has the option to update device and fleet specifications through the Web UI or CLI. The alternative is to keep these specifications in a Git repository and ask the service to periodically load the specifications from that Git repository into its database.

> When the agent receives a new target configuration from the service, the agent does the following:

Reviewer: ```suggestion
When the agent receives a new target configuration from the service, the agent completes the following tasks:
```

. To avoid depending on network connectivity during the update, the agent downloads all required resources, such as the OS image and application container images, over the network to disk.

> . The agent updates the OS image by delegating to `bootc` or `rpm-ostree`.

Reviewer 2: Remove `rpm-ostree`, please. It'll only be supported in a later release.

. The agent updates configuration files on the device's file system by overlaying a set of files that the service sends to the device.
. If necessary, the agent reboots into the new OS.
Otherwise, the agent signals applications to reload the updated configuration.
. The agent updates applications running on Podman or MicroShift.

> If applying any of these changes fails or the system does not return online after rebooting, the agent rolls back to the previous OS image and configuration.

Reviewer: If "applying" fails, or the results fail? 

How does the user know it failed? _Something_ would fail.

_If ___ fails, or the system does not return online, the agent..._

Reviewer 2: If the update fails to apply or the system does not return online after rebooting...

> How does the user know it failed? Something would fail.

The agent knows when it fails to write the changes to disk or when it is supposed to start/stop/update containers on Podman and there's an error there.
It does not know whether the system is in a correctly working state after a reboot. That is what `greenboot` is there for. `greenboot` runs tests defined by RHEL, by RHEM, and by the user and triggers the rollback of the OS update if a required test fails.
The agent can detect that `greenboot` rolled back the OS and will roll back the rest of the config, too.

The user will know, because the agent reports back that the update failed and why.

Reviewer: I am fine with _if the update fails to apply or the system does not return online after rebooting_ . The update is what fails, not the verb: applying.

[#rhem-api-server]
== {rhem} API server

The API server is a core component of the {rhem} service that exposes both agent-facing and user-facing API endpoints.
The {rhem} service communicates with various external systems to authenticate and authorize users, get mTLS certificates signed, or query configuration for managed devices.

> For the external communication, the {rhem} service's API server exposes two endpoints:

Reviewer 2: Redundant with the first sentence ("[...] exposes both agent-facing and user-facing API endpoints"). Also, saying "For the external communication" right after "communicates with various external systems" seems to imply the API endpoints are used with those external systems.

The user- and agent-facing APIs are for the users and agents to connect to the RHEM service, respectively. The RHEM service, in turn, connects to the external systems for configuration, authentication, authorization, and having certificates signed.

> 

Reviewer: Possessive again, will need to rework that line. See the manual for ideas (linked above)

User-facing API endpoint:: Users can connect to the user-facing API endpoint from the CLI or the web UI.

> Users must authenticate with the configured external authentication service to obtain a JSON Web Token (JWT) token with which they can make HTTPS requests.

Reviewer: We don't use UI in the doc, we call it "console"

Agent-facing API endpoint:: Agents connect to the agent-facing endpoint, which is mTLS-protected.

> The service authenticates the device based on its X.509 client certificates. 

Reviewer: same here with "its", can change to "the" I think, without losing meaning.

[#edge-mgmt-arch]

> = Understanding the {rhem} architecture

Reviewer: We don't need to have _understanding_ here -- we can just name this {rhem architecture}

> = Understanding the {rhem} architecture

Reviewer: ```suggestion
{rhem} architecture
```

> You can manage individual or a fleet of devices by using the {rhem}.

Reviewer 3: ```suggestion
You can manage individual devices or an entire fleet using {rhem}.
```

The current wording feels awkward to me.

> The {rhem} uses an agent-based architecture with image-based operating systems (OS), that allows for a scalable and robust device management under limited network conditions.

Reviewer 3: ```suggestion
The {rhem} uses an agent-based architecture with image-based operating systems (OS), that allows for a scalable and robust device management even under limited network conditions.
```

Is "The {rhem}" correct, or should it just be {rhem}?

Author: It's a variable, so when the docs are rendered, it'll spell out `Red Hat Edge Manager` and for that, we should use the definite article. Compare these:
> Red Hat Edge Manager uses...

vs
> The Red Hat Edge Manager uses...

Reviewer 3: I know, I still think it's better without the "the", but I leave it to the expert :)

Reviewer: Also, please make sure we avoid the word "under" in tech doc, in an accessibility test it would be flagged potentially as a direction word that would need justification. You can almost always replace it with a word like _within_ .

The image-based OS that is distributed to the devices includes a {rhem} agent and its configuration.

> This means that a {rhem} agent autonomously handles the management and monitoring of each device while the agent periodically communicates with a {rhem} service to query for new configuration and to report device status.

Reviewer 3: ```suggestion
This means that the {rhem} agent autonomously manages and monitors each device while periodically communicating with the {rhem} service to check for new configurations and report device status.
```

The {rhem} architecture has the following main concepts:

* Agent
* Service
* Image-based OS
* API server 
* Database
* Device
* Device fleet

For more information, see the following documentation:

* <<agent-service,Understanding the {rhem} agent and service>>
* <<api-server,Understanding the {rhem} API server>>
////
TODO include links when docs are ready
- Build images
- Provisioning
- Managing devices
- Managing device fleets
////

[#agent-service]
== Understanding the {rhem} agent and service

The {rhem} agent is a process running on each managed device that periodically communicates with the {rhem} service that runs on the {acm-short} hub cluster.

> During the periodical communication, the agent enrolls the devices into the service, queries the service for changes, and reports the status of devices.

Reviewer 3: ```suggestion
During periodic communication, the agent enrolls devices into the service, checks for updates, and reports device status.
```

When the agent is not actively connected to the service, the agent can apply configuration updates independently from the service.

The {rhem} service handles user and agent authentication and authorization, device enrollment and inventory, rolling out updates to devices, and reporting status from individual devices or fleets.
The service communicates with a database that stores the inventory and runtime information, such as the current target configuration and the reported actual configuration.

> *Note:* You can store the inventory and runtime information in a PostgreQSL database or a Git repository.

Reviewer 3: ```suggestion
*Note:* You can store the inventory and runtime information in a PostgreSQL database or a Git repository.
```

This isn't accurate.  Everything is stored in the PostgreSQL DB (actually some information is also stored in a Redis key-value store).

Users can maintain fleet definitions in Git, and the RHEM service will periodically sync changes from Git to the Fleet definitions in the SQL DB.

Author: Thanks, I misunderstood this. I'll rephrase it

When the agent receives a new target configuration from the service, the agent does the following:

> . To avoid depending on network connectivity during the update, the agent downloads all required resources, such as the OS image and application container images, over the network to disk.

Reviewer 3: @fzdarsky I believe this is the vision but hasn't been implemented

Reviewer 2: Confirmed with @hexfusion that the agent indeed already pulls the OS image and app images before starting the update. There's a small network dependency left in that bootc performs a pull just before initiating the upgrade, but I feel the statement is "true enough".

. The agent updates the OS image by delegating to `bootc` or `rpm-ostree`.
. The agent updates configuration files on the device's file system by overlaying a set of files that the service sends to the device.
. If necessary, the agent reboots into the new OS.
Otherwise, the agent signals services to reload the updated configuration.
. The agent updates applications running on Podman or MicroShift.

If applying any of these changes fails or the system does not return online after rebooting, the agent rolls back to the previous OS image and configuration.

[#api-server]
== Understanding the {rhem} API server

> The API server is a part of the {rhem} service that exposes both agent-facing and user-facing API endpoints.

Reviewer 3: ```suggestion
The API server is a core component of the {rhem} service that exposes both agent-facing and user-facing API endpoints.
```

The {rhem} service communicates with various external systems to authenticate and authorize users, get mTLS certificates signed, or query configuration for managed devices.

For the external communication, the {rhem} service's API server exposes two endpoints:

User-facing API endpoint:: Users can connect to the user-facing API endpoint from the CLI or the web UI.
Users must authenticate with the configured external authentication service to obtain a JSON Web Token (JWT) token with which they can make HTTPS requests.

> Agent-facing API endpoint:: Agents connect to the agent-facing endpoint.
> The endpoint is mTLS-protected which means that the service authenticates the device based on its X.509 client certificates. 

Reviewer 3: ```suggestion
Agent-facing API endpoint:: Agents connect to the agent-facing endpoint, which is protected using mutual TLS (mTLS). The service authenticates each device using its X.509 client certificate.
```

[#edge-mgmt-arch]

> = Understanding the {rhem} architecture

Reviewer 3: I didn't go through the entire section yet because I got a bit lost.  Who is this section targeting and for what purposes?  It reads like random facts about RHEM internals.

Author: @avishayt Could you please take another look? I rewrote parts of it to provide a high-level overview

The {rhem} uses an agent-based architecture, which means that a {rhem} agent runs on each edge device.
The devices periodically communicate with the {rhem} service through the agent.
The agent autonomously handles device management and monitoring.
This agent-based architecture allows for a scalable and robust device management under limited network conditions.

The {rhem} architecture has the following main concepts:

Agent:: The {rhem} agent runs on each device. The agent periodically connects to the {rhem} service for monitoring and updates.
Service:: The {rhem} service handles user and agent authentication and authorization, device enrollment and inventory, rolling out updates to devices, and rolling up status from devices.
API server:: A part of the {rhem} service that exposes both agent-facing and user-facing API endpoints.
Device:: A combination of a real or virtual machine, an operating system, and application workloads that function together to serve a specific purpose.

> Device spec:: A specification of a state of the device that the user defines.

Reviewer 3: We generally call this the "desired state".

Device status:: A record of the state that the device is reported to actually have.
Device template:: A template for device specs that serves to control deviations between the configurations of devices.
Fleet:: A group of devices governed by a common device template and common management policies.
PostgreSQL database:: A database for storing inventory and runtime information, such as the current target configuration and the reported actual configuration.
////
Field Selector:: Filters and selectors for {rhem} objects based on the values of specific resource fields.

> Field selectors follow the same syntax, principles, and support the same operators as Kubernetes field and label selectors.

Reviewer 3: I don't know if we need to get into this here, but we actually support a superset of Kubernetes.

Labels:: Organizers for devices and other resources.
Label Selector:: Groups or filters for devices and other resources based on assigned labels.

> ////

Author: I think the concepts related to the organization of devices should be in "Organizing devices" or similar. I commented them out for now to allow for comments in case there are any objections.

[#edge-mgmt-arch]
= Understanding the {rhem} architecture

The {rhem} uses an agent-based architecture, which means that a {rhem} agent runs on each edge device that periodically communicates with the {rhem} service for device management and monitoring.
The agent autonomously updates the device and reports the device status to the service.
This agent-based architecture allows for a scalable and robust device management under limited network conditions.

The {rhem} service consists of the following components:

* an API server that exposes both agent-facing and user-facing API endpoints
* a PostgreSQL database for storing inventory and runtime information such as the current target configuration and the reported actual configuration
* worker processes

////
I don't think it makes sense to include this whole list of concepts here. Agent, Service, Device, Device fleet, and template might fit here. However, label and field selectors should be covered in "Organizing devices" or similar
Agent:: The {rhem} agent runs on each device. The agent periodically connects to the {rhem} service for monitoring and updates.

> Service:: The {rhem} service handles user and agent authentication and authorization, device enrollment and inventory, rolling out updates to devices, and rolling up status from devices.

Author: What is "rolling up status from devices"? Is it status monitoring?

Reviewer 3: It can report the status of individual devices as well as aggregate the status of multiple devices.

//What is "rolling up status from devices"? Is it status monitoring basically?
Device:: A combination of a real or virtual machine, an operating system, and application workloads that function together to serve a specific purpose.
Device Spec:: A specification of a state of the device that the user defines.
Device Status:: A record of the state that the device is reported to actually have.
Device Template:: A template for device specs that serves to control deviations between the configurations of devices.
Fleet:: A group of devices governed by a common device template and common management policies.
Field Selector:: Filters and selectors for {rhem} objects based on the values of specific resource fields.
Field selectors follow the same syntax, principles, and support the same operators as Kubernetes field and label selectors.
Labels:: Organizers for devices and other resources.

> Label Selector:: Groups or filters for devices and other resources based on assigned labels.

Reviewer 4: I don't think it makes sense to include this whole list of concepts here. Agent, Service, Device, Device fleet, and template might fit here. However, label and field selectors should be covered in "Organizing devices" or similar

////

[#edge-mgmt-api-server]
== Understanding the {rhem} API server

The {rhem} service communicates with various external systems to authenticate and authorize users, get mTLS certificates signed, or query configuration for managed devices.

For the external communication, the {rhem} service's API server exposes two endpoints:

User-facing API endpoint:: Users can connect to the user-facing API endpoint from the CLI or the web UI.
Users must authenticate with the configured external authentication service to obtain a JSON Web Token (JWT) token with which they can make HTTPS requests.

Agent-facing API endpoint:: Agents connect to the agent-facing endpoint.

> The endpoint is mTLS-protected which means that the service authenticates the device based on its X.509 client certificates. 

Author: "its X.509 client certificates"

Is it the service's or the device's cert?

Reviewer 2: the device's client cert

The {rhem} agent is a process running on each managed device that periodically communicates with the service.
The periodic reporting allows the device to be on a private network or to have a dynamic IP address.

The agent is responsible for the following processes:

* securely enrolling devices with the specified target configuration
* periodically monitoring for new configuration from the service
* autonomously applying updates
* periodically notifying the service if a configuration deviates from the target configuration
* reporting status on the updates and the health of devices back to the service

Thanks to the declarative nature of the device configuration, you can store the target configuration in a Git repository.
The {rhem} service can periodically poll the repository for updates or receive updates from a webhook.

When the agent receives a new target configuration from the service, the agent does the following:

. To avoid depending on network connectivity during the update, the agent downloads all required resources, such as the OS image and application container images, over the network to disk.
. The agent updates the OS image by delegating to `bootc` or `rpm-ostree`.

> . The agent updates configuration files on the device's file system by overlaying a set of files that the service sends to the device.

Author: What do you mean by "overlaying a set of files"? Is it overwriting or is it different than that?

Reviewer 3: "Overlaying" is a common term for dealing with file systems.  The agent will create new files when necessary, and overwrite files when necessary.

Reviewer: Possessive in line 46 we need to change around:

https://www.ibm.com/docs/en/ibm-style?topic=grammar-possessives

//What do you mean by "overlaying a set of files"? Is it overwriting or is it different than that?
. If necessary, the agent reboots into the new OS.

> Otherwise, the agent signals services to reload the updated configuration.

Author: If we use service for the RHEM service and services here, it can get confusing. Could we use workloads or something else instead?

Reviewer 3: @fzdarsky I think "applications" would fit our terminology best?

Reviewer 2: I agree just "services" is confusing.
"Workloads" would refer to anything running on the system and producing "load", i.e. both the user's "applications" as well as system services (systemd) are workloads.
The sentence was referring to workloads, but I'm wondering how accurate we need to be here. We could also say "[...] signals to system services and applications to [...]"
